### PR TITLE
W-16823349

### PR DESCRIPTION
**Pull request by IvanAndresFritzler**

### CODE REVIEW

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/spring/ByteBuddySpringCacheInstrumentator.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+/**
+ * This class offers instrumentation alternatives for Spring {@link org.springframework.util.ConcurrentReferenceHashMap} caches
+ * cleanup. Spring caches of the type {@link org.springframework.util.ConcurrentReferenceHashMap} are scattered all over the
+ * Spring framework code. Those caches are never explicitly cleaned up (the cleanup relies on the values being soft referenced)
+ * and can cause mule artifact classloaders retention.
+ */
+public class ByteBuddySpringCacheInstrumentator {
+
+  private static final String CONCURRENT_REFERENCE_HASH_MAP_CLASS = "org.mule.springframework.util.ConcurrentReferenceHashMap";
```
Comment (eze210) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/spring/ByteBuddySpringCacheInstrumentator.java:33]: Clarify that the `.mule.` part comes from the shading

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/spring/ByteBuddySpringCacheInstrumentator.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    TypePool typePool = TypePool.Default.of(targetClassloader);
+    try (DynamicType.Unloaded<?> unloaded = new ByteBuddy()
+        .redefine(typePool.describe(CONCURRENT_REFERENCE_HASH_MAP_CLASS).resolve(),
+                  ClassFileLocator.ForClassLoader.of(IntrospectionUtils.class.getClassLoader()))
+        // We redirect any other constructor to the fully parameterized one.
+        .constructor(target -> target.isConstructor() && target.getParameters().size() < 4)
+        .intercept(MethodCall
+            .invoke((ElementMatcher<MethodDescription>) target -> target.isConstructor() && target.getParameters().size() == 4)
+            // We use default arguments plus WEAK reference type, discarding the intercepted ones.
+            .with(16, 0.75F, 16, WEAK))
```
Comment (eze210) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/spring/ByteBuddySpringCacheInstrumentator.java:87]: Just curiosity: Are you going to use the provided values or these hardcoded ones?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/spring/ByteBuddySpringCacheInstrumentator.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        .make()) {
+      unloaded.load(IntrospectionUtils.class.getClassLoader(), ClassLoadingStrategy.Default.INJECTION);
+      LOGGER.debug("Spring caches are now instrumented for weakness.");
```
Comment (eze210) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/spring/ByteBuddySpringCacheInstrumentator.java:90]: `for weakness` sounds strange. Maybe `for using weak references`

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/spring/ByteBuddySpringCachesManager.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import org.springframework.core.SpringVersion;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Optional.empty;
+import static org.slf4j.LoggerFactory.getLogger;
+
+/**
+ * ByteBuddy aspect implementation that stores Soring {@link org.springframework.util.ConcurrentReferenceHashMap} caches, offering
```
Comment (eze210) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/spring/ByteBuddySpringCachesManager.java:22]: Soring

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/spring/ByteBuddySpringCachesManager.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ * a cleanup method. The use the cleanup feature, this aspect must be instrumented through
+ * {@link ByteBuddySpringCacheInstrumentator#instrumentForCleanup(ClassLoader)} call.
+ * 
+ * @see ByteBuddySpringCacheInstrumentator#instrumentForCleanup(ClassLoader)
+ */
+public class ByteBuddySpringCachesManager {
+
+  private static final Logger LOGGER = getLogger(ByteBuddySpringCachesManager.class);
+  public static final Cache<Object, Optional<?>> springCaches = Caffeine.newBuilder().weakKeys().build();
```
Comment (eze210) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/spring/ByteBuddySpringCachesManager.java:31]: Using weakKeys in this cache could make the clearCaches to "forget" to clean some cache. Wouldn't a plain old ConcurrentHashMap work?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/IntrospectionUtils.java [hunk 6]
```diff
@@ -183,33 +182,13 @@ public final class IntrospectionUtils {
   private static final Logger LOGGER = getLogger(IntrospectionUtils.class);
 
   static {
-    setWeakHashCaches();
+    // A shade of the spring-core library is the cause for this additional cleanup.
```
Comment (eze210) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/util/IntrospectionUtils.java:185]: Please add a comment referencing this in the pom.xml that configures the shade plugin, so we don't forget to remove this in that case.

**Review by eze210 — DISMISSED** [review]

**Review by fsgonz — APPROVED** [review]

**Pull request merged**